cpp/include/cudf/stream_compaction.hpp

```diff
@@ -24,55 +24,82 @@
+ * @param[in] input The input table to filter
  * @param[in] boolean_mask A column of type GDF_BOOL8 used as a mask to filter
  * the input column corresponding index passes the filter.
-  * @return gdf_column Column containing copy of all elements of @p input passing
+  * @return cudf::table Table containing copy of all rows of @p input passing
  * the filter defined by @p boolean_mask.
  */
-gdf_column apply_boolean_mask(gdf_column const &input,
-                              gdf_column const &boolean_mask);
+table apply_boolean_mask(table const &input,
+                         gdf_column const &boolean_mask);
 
 /**
- * @brief Filters a column to remove null elements.
- *
- * Given an input column an element `i` from the input column is copied to the
- * output if the corresponding element `i` in the input's valid bitmask is
- * non-null.
+ * @brief Enumeration used to indicate ALL true/valid or ANY true/valid
+ * 
+ * Used by, for example, cudf::drop_nulls().
+ **/
+enum any_or_all {
+  ANY = 0,
+  ALL
+};
+
+/**
+ * @brief Filters a table to remove null elements.
  *
- * The output column has size equal to the number of elements in boolean_mask 
- * that are both non-null and `true`. Note that the output column memory is 
- * allocated by this function but must be freed by the caller when finished.
+ * Filters the rows of the input table considering only specified columns for
+ * validity / null values.
  * 
- * If the input column is not nullable, this function just copies the input
- * to the output.
+ * Given an input table, row `i` from the input columns is copied to the
+ * output if the row is not null. Null means:
+ *  - If @p drop_if is ANY, that there is a null in any column of @p keys at
+ *    that row.
+ *  - If @p drop_if is ALL, that there are nulls in all columns of @p keys at
+ *    that row.
+ *
+ * This operation is stable: the input order is preserved in the output.
  * 
- * * @note if @p input.size is zero, there is no error, and an empty column is 
- * returned.
+ * Note that the memory for the columns of the output table is allocated by this
+ * function but must be freed by the caller when finished.
+ *
+ * Any non-nullable column in the input is treated as all non-null.
+ *
+ * @note if @p input.num_rows() is zero, or @p keys is empty or has no nulls,
+ * there is no error, and an empty table is returned
  * 
- * @param[in] input The input column to filter
- * @return gdf_column Column containing copy of all non-null elements of @p input.
+ * @throws cudf::logic_error if @p keys is non-empty and keys.num_rows() is less
+ * than input.num_rows()
+ *
+ * @param[in] input The input table to filter.
+ * @param[in] key The table of columns to check for nulls.
+ * @param[in] drop_if If ANY, drop rows that have a null in any column.
+ *                    If ALL, drop rows that have a null in all columns.
+ * @param[in] valid_threshold Required number of non-null fields in a row to
```

Comment:
With the addition of `valid_threshold`, the `drop_if` parameter seems somewhat redundant, as it can be emulated by passing in a `valid_threshold` of `num_columns` or `0`, for `ANY` and `ALL` respectively.

Reply:
Yeah, I was just mirroring Pandas, which is probably not a great idea.

Actually the drop_if semantics are "ANY = drop if any are null" and "ALL = drop if all are null", and the threshold is a keep threshold, not a drop threshold. SO I believe it's:

`ANY => threshold = num_columns`
`ALL => threshold = 0`

The problem is I want to make threshold a default parameter, but I want the default semantics to be to drop if there's ANY nulls (like Pandas). That would require a NULL threshold of num_columns, which is not possible to default.  So I'm thinking of flipping it to be `null_threshold`, which would have the semantics you describe, but the opposite of Pandas...

@shwina do you think it would be bad to have semantics opposite to Pandas here?

Reply:
The `null_threshold` seems like a reasonable idea.

Though, you can work around the default argument issue like this:
```
drop_nulls(table const &input, table const &keys, gdf_size_type valid_threshold);

drop_nulls(table const &input, table const &keys){
   return drop_nulls(input, keys, input.size());
}
```

Reply:
OK, now there are two versions as you suggest, with a `keep_threshold` parameter that has the same semantics as Pandas, but without the redundancy of any/all. I left the python `apply_drop_nulls` API alone, so it matches pandas.